### PR TITLE
[CES-157] - Implement tooling for storage account migration from WEU to ITN

### PR DESCRIPTION
**Pull request by cfcuffari120400**
### Motivation and Context

The decided strategy for the migration of storage accounts to Italy North entails the usage of Data Factory.
This PR introduces a module and scaffolding to create the necessary resources.

### Major Changes

Create a module and scaffolding for the creation of data factory resources for the copy of blob and table data in storage accounts.

### Testing

The resources have been tested deeply in the DEV-ENGINEERING subscription.
The following is an example of configuration derived from the migration of a test storage account:
Example of data factory configuration used for tests:
image

### CODE REVIEW

**Review by gunzip — CHANGES_REQUESTED** [review]
I'd do the following:
- link a Jira issue into the title (ie. [CES-100]
- use a more descriptive title
- follow pull request template describing the PR

Feel free to keep PRs in draft until review is needed.

File: src/_modules/test_users/DataFactory/environments/dev/main.tf [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+module "data_factory" {
+  source = "../../modules/data_factory"
+
+  data_factories            = var.data_factories
+  pipelines                 = var.pipelines
+  azure_runtimes            = var.azure_runtimes
+  self_hosted_runtimes      = var.self_hosted_runtimes
+  linked_services           = var.linked_services
+  datasets                  = var.datasets
+  storage_account_resource_groups = var.storage_account_resource_groups  # Aggiungi questa riga
```
Comment (gunzip) [inline src/_modules/test_users/DataFactory/environments/dev/main.tf:18]:
```suggestion
  storage_account_resource_groups = var.storage_account_resource_groups
```

**Review by gunzip — CHANGES_REQUESTED** [review]
Keep in mind that everything that is committed must be in english.

**Review by mamu0 — CHANGES_REQUESTED** [review]
The directory structure is incorrect because the `data_factory` module should be placed within the `_modules` folder, at the same level as `test_users`. Additionally, the `environment/dev` path should be placed within the relevant domain where the resource is intended to be created.

Also, the `dev` folder content needs review, as the final configuration for production should already be defined in this repo

File: src/migration/prod/locals.tf [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+  tags = {
+    CostCenter  = "TS310 - PAGAMENTI & SERVIZI"
+    CreatedBy   = "Terraform"
+    Environment = "Prod"
+    Owner       = "IO"
+    Source      = "https://github.com/pagopa/io-infra/blob/main/src/migration/prod"
+  }
+
+  storage_accounts = [
```
Comment (mamu0) [inline src/migration/prod/locals.tf:27]: should this list be updated once the storage accounts in itn have been created?

File: src/migration/prod/italynorth.tf [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  tags = local.tags
+}
+
+# Create Azure Data Factory instances
+# Enables system-assigned managed identity for secure access to resources
+resource "azurerm_data_factory" "this" {
+  name                = "${local.project_itn}-${local.environment.domain}-${local.environment.app_name}-adf-${local.environment.instance_number}"
+  location            = "italynorth"
+  resource_group_name = azurerm_resource_group.migration.name
+
```
Comment (mamu0) [inline src/migration/prod/italynorth.tf:14]: Now is allowed public access as default, maybe you can add
`public_network_enabled = false`

File: src/migration/prod/locals.tf (hunk 2 above)
Comment (christian-calabrese) [reply src/migration/prod/locals.tf:27]: Yes, exactly! Or we can decide to instantiate the module close to the storage accounts to be migrated

File: src/migration/prod/italynorth.tf (hunk 3 above)
Comment (christian-calabrese) [reply src/migration/prod/italynorth.tf:14]: Yes, I need to understand whether it needs some other resources to work privately

File: src/_modules/data_factory_storage_account/data.tf [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+  name                = var.storage_accounts.target.name
+  resource_group_name = var.storage_accounts.target.resource_group_name
+}
+
+data "azurerm_storage_containers" "this" {
+  count              = var.what_to_migrate.blob.enabled && length(var.what_to_migrate.blob.containers) == 0 ? 1 : 0
+  storage_account_id = data.azurerm_storage_account.source.id
+}
+
+data "azapi_resource_list" "tables" {
```
Comment (Krusty93) [inline src/_modules/data_factory_storage_account/data.tf:16]: Why this provider? Azurerm includes table storage

Comment (christian-calabrese) [reply src/_modules/data_factory_storage_account/data.tf:16]: I couldn't find a better way to list the tables in a storage account with the azurerm provider.
This way, it is not needed to specify each table name to migrate. The module will migrate all of them straight away

**Review by mamu0 — APPROVED** [review]
LGTM

**Comment by gunzip** [comment]
please use the pull requets template and provide motivation and context

**Comment by christian-calabrese** [comment]
> please use the pull requets template and provide motivation and context.
> 
> PR title should reflect purpose, not changes.

PR title and description updated accordingly. Does it look ok?

**Review by christian-calabrese — APPROVED** [review]

**Pull request merged**